native/src/main/java/io/ballerina/lib/avro/deserialize/visitor/DeserializeVisitor.java

```diff
@@ -83,7 +83,8 @@ public BMap<BString, Object> visit(RecordDeserializer recordDeserializer, Generi
         Type originalType = recordDeserializer.getType();
         Type type = recordDeserializer.getType();
         Schema schema = recordDeserializer.getSchema();
-        BMap<BString, Object> avroRecord = createAvroRecord(type);
+        BMap<BString, Object> avroRecord = (type.getTag() == TypeTags.ANYDATA_TAG)
+                ? ValueCreator.createMapValue() : createAvroRecord(type);
```

Comment:
Can't we create a map value passing the type, here?

Reply:
There are some complex types like `record{} & readonly`, where we cannot use the `BMap` value directly. Because value creation fails for `readonly` type.